File: ios/ReactNativeTemplate/RNUtilModule.m
#import "RNUtilModule.h"
#import <AppTrackingTransparency/AppTrackingTransparency.h>
#import <AdSupport/ASIdentifierManager.h>

@implementation RNUtilModule

RCT_EXPORT_MODULE();

RCT_REMAP_METHOD(getAdvertisingStatus, getAdvertisingStatus_resolver: (RCTPromiseResolveBlock) resolve getAdvertisingStatus_rejecter: (RCTPromiseRejectBlock) reject)
{
  if (@available(iOS 14, *)) {
    resolve([NSNumber numberWithLong:[ATTrackingManager trackingAuthorizationStatus]]);
  } else {
    resolve([NSNumber numberWithInt:3]);
  }
}

RCT_REMAP_METHOD(getAdvertisingInfo, getAdvertisingInfo_resolver: (RCTPromiseResolveBlock) resolve getAdvertisingInfo_rejecter: (RCTPromiseRejectBlock) reject)
{
  NSMutableDictionary *result = [[NSMutableDictionary alloc] init];
  [result setObject:[NSNull null] forKey:@"id"];
  [result setObject:[NSNumber numberWithBool:true] forKey:@"isAdTrackingLimited"];

  if (@available(iOS 14, *)) {
    switch ([ATTrackingManager trackingAuthorizationStatus]) {
      case ATTrackingManagerAuthorizationStatusAuthorized:
        [result setObject:[[[ASIdentifierManager sharedManager] advertisingIdentifier] UUIDString] forKey:@"id"];
        [result setObject:[NSNumber numberWithBool:false] forKey:@"isAdTrackingLimited"];
        resolve(result);
        break;
      case ATTrackingManagerAuthorizationStatusRestricted:
      case ATTrackingManagerAuthorizationStatusDenied:
        resolve(result);
        break;
      case ATTrackingManagerAuthorizationStatusNotDetermined:
        [ATTrackingManager requestTrackingAuthorizationWithCompletionHandler:^(ATTrackingManagerAuthorizationStatus status) {
          ATTrackingManagerAuthorizationStatus currentStatus = [ATTrackingManager trackingAuthorizationStatus];
          while (currentStatus == ATTrackingManagerAuthorizationStatusNotDetermined) {
            currentStatus = [ATTrackingManager trackingAuthorizationStatus];
            [NSThread sleepForTimeInterval:0.1f];
          }

          switch (currentStatus) {
            case ATTrackingManagerAuthorizationStatusAuthorized:
              [result setObject:[[[ASIdentifierManager sharedManager] advertisingIdentifier] UUIDString] forKey:@"id"];
              [result setObject:[NSNumber numberWithBool:false] forKey:@"isAdTrackingLimited"];
              resolve(result);
              break;
            default:
              resolve(result);
              break;
          }
        }];
        break;
    }
  } else {
    [result setObject:[[[ASIdentifierManager sharedManager] advertisingIdentifier] UUIDString] forKey:@"id"];
    [result setObject:[NSNumber numberWithBool:false] forKey:@"isAdTrackingLimited"];
    resolve(result);
  }
}

@end

File: ios/ReactNativeTemplate/RNUtilModule.swift
import Foundation
import AppTrackingTransparency
import AdSupport
import React // 추가: RCTPromiseResolveBlock, RCTPromiseRejectBlock 사용을 위해 필요

@objc(RNUtilModule)
class RNUtilModule: NSObject {
  
  @objc
  func getAdvertisingStatus(_ resolve: @escaping RCTPromiseResolveBlock,
                            rejecter reject: @escaping RCTPromiseRejectBlock) {
    if #available(iOS 14, *) {
      resolve(NSNumber(value: ATTrackingManager.trackingAuthorizationStatus.rawValue))
    } else {
      resolve(NSNumber(value: 3)) // iOS 14 미만에서는 authorized로 간주
    }
  }
  
  @objc
  func getAdvertisingInfo(_ resolve: @escaping RCTPromiseResolveBlock,
                          rejecter reject: @escaping RCTPromiseRejectBlock) {
    var result: [String: Any] = [
      "id": NSNull(),
      "isAdTrackingLimited": true
    ]
    
    if #available(iOS 14, *) {
      switch ATTrackingManager.trackingAuthorizationStatus {
      case .authorized:
        let idfa = ASIdentifierManager.shared().advertisingIdentifier.uuidString
        result["id"] = idfa
        result["isAdTrackingLimited"] = false
        resolve(result)
      case .restricted, .denied:
        resolve(result)
      case .notDetermined:
        ATTrackingManager.requestTrackingAuthorization { status in
          let currentStatus = ATTrackingManager.trackingAuthorizationStatus
          if currentStatus == .authorized {
            let idfa = ASIdentifierManager.shared().advertisingIdentifier.uuidString
            result["id"] = idfa
            result["isAdTrackingLimited"] = false
          }
          resolve(result)
        }
      @unknown default:
        resolve(result)
      }
    } else {
      let idfa = ASIdentifierManager.shared().advertisingIdentifier.uuidString
      result["id"] = idfa
      result["isAdTrackingLimited"] = false
      resolve(result)
    }
  }
}
